### ISSUE TITLE
Add Apply between Monad and Functor

### ISSUE THREAD

**Issue by jbgi** [javaslang/javaslang#240]
Applicative functors are extremely useful, it would help a lot to have Apply for implementing lenses.
Also not sure why Traversable extends Monad ???
Also Monad and Functor are generated to avoid some duplication with the 'Checked' alternative ; but I think it it is probably not worth it ; hacking them (to add default methods for example) is extremely tedious. WDYT?

**Comment by danieldietrich** [comment]
In Javaslang Tranversable is a representation of recursive 'foldable' data types, and it has `map()`, `flatMap()`etc. Of cause that does not make Traversable a Functor or a Monad. Actually all implementing classes (List and Stream) are Monads, so it was a matter of object oriented refactoring to generalize the similarities and group them in a super class/interface. This is all about it - no special reason.

But we can do it the right way - what would you suggest? Instead of `Apply` I like the full name `Applicative`. I do not like most of the naming in `functionaljava`. Perhaps because I'm more a 'Scala' than a 'Haskell'...

Make a suggestion! Pull requests are welcome! :-)

**Comment by danieldietrich** [comment]
Ahh, sorry, I forgot to answer your question about code generation. Please try to generate duplicate code (minimum granularity: classes), otherwise it would be only a matter of time that code / javadoc diverges. I would like to assist you, if you need help!

**Comment by danieldietrich** [comment]
I prototype code always outside of a generator (just as usual), copy it (or parts of it) into `generator/Generator.scala` and create generator templates.

You may commit code as-is. I will do the generator-stuff if necessary. If you need to change Functor/CheckedFunctor and Monad/CheckMonad you may copy them from `src-gen` to `src` and comment out these lines in Generator.scala:

```scala
genMatch()
genFunctions()
// genFunctor()
// genMonad()
genPropertyChecks()
genTuples()
```

**Comment by danieldietrich** [comment]
@jbgi /cc @mperry @lukaseder

The algebraic types in Javaslang are mainly used as an API design aspect to encapsulate common methods and mark/declare specific properties of classes. We have {Functor + Monad}, extended by many types in Javaslang, and Monoid, used only in a fold-Method of Traversables.

I've got recent feedback from users that they are 'in fear' of the algebra package.

Please tell me more about the things you want to implement and how they integrate with the existing classes - or should they be used on their own?

My main goal for Javaslang is to create a Java extension which takes the most out of the new Java features and to bring immutability _as default_ to Java. I target the majority of Java developers out there and I'm sure that many are even not familiar in depth with variance types regarding generics. I think that sophisticated algebraic stuff would keep users away.

I'm really excited that you bring more functional mojo to Javaslang. And I'm sure it is useful. Maybe a module javaslang-algebra would make sense for the users, who want to go really deep down the rabbit hole?

**Comment by danieldietrich** [comment]
More specific I see Javaslang as a simple Scala. The algebraic extension is more like Scalaz.

**Comment by danieldietrich** [comment]
I good name for the module would be `javaslang-pure`! :-)

**Comment by lukaseder** [comment]
> My main goal for Javaslang is to create a Java extension which takes the most out of the new Java features and to bring immutability as default to Java. I target the majority of Java developers out there and I'm sure that many are even not familiar in depth with variance types regarding generics. I think that sophisticated algebraic stuff would keep users away.

I'm going to be curious how you'll pull that off:

- Without declaration-site variance
- Without causing headaches at the call site via use-site variance

I've seen your trick involving recursive generics in `Kind`. If this is unleashed upon your users, they'll run :-)

**Comment by danieldietrich** [comment]
@lukaseder You misunderstood, of cause Java classes with generic type parameters are invariant. However, Arrays are covariant, overriding methods are covariant in the return type, and - what I meant, upper and lower bounds on generic type parameters can also be used in this context. Many Java 'experts' are not aware of these or use them not correctly.

Let's not go to deep in a technical discussion here. Let's think about how we can bring funky algebra trickery to the `javaslang-pure` module! :-)

**Comment by danieldietrich** [comment]
@lukaseder 

> I've seen your trick involving recursive generics in Kind. If this is unleashed upon your users, they'll run :-)

I hope they will stay with me.

Maybe I invent a new paradigm 'Kind Oriented Programming' KOP&trade;

Of course there will be also a 'Kindly Manifesto'.

#ICanHazKind :-)

**Comment by jbgi** [comment]
@danieldietrich ok. now what I would suggest is to refrain from using algebra/category theory terms in javaslang(-core) unless the actual abstraction strictly adhere to the associated definition/laws. And we will let the pure module provides the correct implementations of those abstractions, (alla highj).
So if, for instance, you want to make Monad extend Iterable then let's not use Monad as a name, please rename it to WithFlatMap or something. (and maybe actually rename the algebra package to internal)
Why? because otherwise it will confuse people that are trying to learn functional programming abstractions. They will have the wrong idea of what Monad or other FP abstractions are. (it's already confusing that the current Monad interface does not provide the unit method).

**Comment by danieldietrich** [comment]
Yes, it has to evolve a little bit. Monad will not extend Iterable, that was no good idea. The javaslang.algebra classes should move to javaslang-pure (instead of renaming the package _internal_). My idea is, to make javaslang-pure the core of javaslang. The javaslang module should be kept ad stable as possible. The foundation may evolve over time.

I want to take part in the design of the core, e.g. with code reviews, talk about conventions like naming, packages, etc.

We will get this to the next Javaslang level! This will be awesome!

**Comment by lukaseder** [comment]
> I hope they will stay with me.

Well, as long as you can hide the generics from client code, you'll be fine. Type inference and type hierarchies are your friends :)

**Comment by danieldietrich** [comment]
@lukaseder I'm relatively satisfied with the current API. The generics of Traversable and Seq are distracting so far.

**Comment by jbgi** [comment]
IMO collection views and conversion methods is a better alternative to a complex type hierarchies that often cause blurring of concepts. It also allows to get ride of (confusing) variance annotations.

**Comment by danieldietrich** [comment]
Sounds interesting, to investigate it, we need a technical spike/prototype in a branch.

The javaslang-pure package will probably contain only interfaces of algebraic types + unit tests.

**Comment by danieldietrich** [comment]
The javaslang-pure package will probably contain only interfaces of algebraic types (not _data_ types) + unit tests.

**Comment by lukaseder** [comment]
> IMO collection views and conversion methods is a better alternative to a complex type hierarchies that often cause blurring of concepts. It also allows to get ride of (confusing) variance annotations.

I can second that. There is almost nothing in jOOQ's (public) type hierarchy that I don't somehow regret 6 years later. Conversion methods are much easier and intuitive to handle and use at the call site.

And if I've ever regretted something *most*, then it's using recursive generics (without declaration-site variance) :)

**Comment by danieldietrich** [comment]
I have to see examples based on Javaslang/how to do it better and see how the API 'feels' from the call site.

Because I'm really deeply convinced that the current API makes so much sense.

I'm thankful about every gist showcase here...

**Comment by danieldietrich** [comment]
@jbgi @lukaseder what are the main pain points of the new tree api #256 in your opinion.

It is simplified there. The Kind arguments `Tree<M, T>` will be also there. How would you remove the `<M>`?

What happens to the flatMap methods (which need the Kind)?

How could views and conversion methods help here?

Are there other issues using interfaces here in your opinion (beside security aspects)?

Your answers would help a lot to get this in the right direction!!

**Comment by lukaseder** [comment]
> It is simplified there. The Kind arguments Tree<M, T> will be also there.

Without actually having tried the API, I suspect that the biggest pain point will be that it is very hard to assign a reference to a variable of the super type `Tree` short of resorting to

- Raw types
- Wildcards

E.g.

```java
Tree screwIt = ...
Tree<?, ?> dontCare = ...
```

In fact, things get worse if someone wants to put `Tree` in a container, as then, there is no way to capture the correct type even with wildcards:

```
Option<Tree<?, ?>> wtfDoesntCompile = ...
```

And you might even risk running in a very rough edge case of Java generics, where type assignments are undecidable (see #9 in this list: http://blog.jooq.org/2014/11/03/10-things-you-didnt-know-about-java, or better, read Ross Tate's full paper on Java wildcards: http://www.cs.cornell.edu/~ross/publications/tamewild/tamewild-tate-pldi11.pdf)

Again, my friendly advice is to stay clear of recursive generics. You'll only get hurt :) I've tried to summarise this in the following article:
http://blog.jooq.org/2013/06/28/the-dangers-of-correlating-subtype-polymorphism-with-generic-polymorphism/

TL;DR: If you correlate generic and subtype polymorphism, the only useful types at the use-site will be the (final) leaf types. All of the other types in the hierarchy are useful only for your internals. That's probably not what you want when defining a useful type like `Tree`

>  How would you remove the <M>?

Hard to say, that's your job :)

> What happens to the flatMap methods (which need the Kind)?

I think you'll never find a good solution prior to Java 10, *if* declaration-site variance makes it into Java.

> How could views and conversion methods help here?

Not sure.

> Are there other issues using interfaces here in your opinion (beside security aspects)?

What do you mean?

**Comment by danieldietrich** [comment]
@lukaseder Thank you, Lukas! Also for the links you provided - very insightful, this takes me to the ground. Wherefrom do you take the time to manage all this? :-)

The recursive type parameter will not leak into the next release. These flatMap() and flatten() functions really tip me over the edge. Need a beer.

**Comment by lukaseder** [comment]
> Wherefrom do you take the time to manage all this? :-)

I have an evil twin.

> Need a beer.

That's the cause *and* solution of all problems.

**Comment by danieldietrich** [comment]
Everyone should have a Lukas :-)

Late night coffee is also hard to handle...

**Comment by danieldietrich** [comment]
@jbgi [The semigroupoids package](http://hackage.haskell.org/package/semigroupoids) of Edward Kmett would be a great reference for javaslang-pure - I think highj uses it also...

**Comment by jbgi** [comment]
@danieldietrich true!

**Comment by danieldietrich** [comment]
@lukaseder @jbgi 

I've thought about all possibilities to...

- remove recursive generic parameters
- have a consistent (monadic) type hierarchy
- get flatMap() right, e.g. List(Some(1), None).flatMap(e -> e) = List(1)

The result is, that I had to remove the whole `algebra` package, which was not in the right place within this project.

The only loss is, that we now have no common interface for monads (beside other minor changes).

@jbgi Now we have to think about how to integrate javaslang-pure - or may it exist on its own?

Lukas, thanks for pushing me!

**Comment by jbgi** [comment]
@danieldietrich for a module similar to semigroupoids to exists there is only one requirement: that all public data and function classes, that are leaf in the type hierarchy, implement Kind (it does not need to appear elsewhere like in methods signature). But without Kind no semigroupoids/pure module...

**Comment by danieldietrich** [comment]
@jbgi 

We could do that only in a flat way:

```java
interface List<T> extends Seq<T>, Kind<List<?>, T> {}
interface Stream<T> extends Seq<T>, Kind<Stream<?>, T> {}
interface Option<T> extends Kind<Option<?>, T> {}
interface Try<T> extends Kind<Try<?>, T> {}
interface Either.LeftProjection<L, R> extends Kind<LeftProjection<?>, T> {}
interface Either.RightProjection<L, R> extends Kind<RightProjection<?>, T> {}
interface Match<T> extends Kind<Match<?>, T> {}
interface Gen<T> extends Kind<Gen<?>, T> {}
interface Arbitrary<T> extends Kind<Arbitrary <?>, T> {}
```

No inheritance possible then with a new kind of `Kind`, e.g. this will then not work:

```java
interface ArrayList<T> extends List<T>, Kind<ArrayList<?>, T> {}
```

**Comment by danieldietrich** [comment]
Also all trees are then of one `Kind`. Maybe this is a solution until a future Java version...

**Comment by danieldietrich** [comment]
@jbgi Possibly a proxy-mechanism is possible which also augments children (linked list tails and tree children):

```java
  Traversable
       |
     Tree<T> <-------- TreeKind<T> extends Tree<T>, Kind<Tree<?>, T>
 (Nil, Branch)=final
       |
 BinaryTree<T> <------ BinaryTreeKind<T> extends BinaryTree<T>, Kind<BinaryTree<?>, T>
 (Nil, Branch)=final
       |
RedBlackTree<T> <----- RedBlackTreeKind<T> extends RedBlackTree<T>, Kind<RedBlackTree<?>, T>
 (Nil, Branch)=final
```

But looking at this, it seems to blow up all too much and it is adding too much _trickery_

**Comment by lukaseder** [comment]
Oh, you've removed the recursive generics already from the API? It is still a very interesting application of generics. I'd love for you to endeavour to do more exploration and learn from you. Just wanted to emit a reminder from someone maintaining a 6-year-old API with 1-2 regrets :)

**Comment by danieldietrich** [comment]
There where several caveats with that. Monadic operations have to be relaxed to be practicable. Scala does this also in Traversable. From the end-user API perspective there is nothing missing. From the API design perspective the algebraic interfaces passed away. They where used technically only somehow. I think in a production-ready real world application no one really

**Comment by danieldietrich** [comment]
@lukaseder I'm thankful, this was the right solution.

There where several caveats with that. Monadic operations have to be relaxed to be practicable. Scala does this also in Traversable. From the end-user API perspective there is nothing missing. From the API design perspective the algebraic interfaces passed away. They where used technically only somehow.

In my local fork `Kind` exists and I currently experiment how to add value for users of Javaslang, without mystifying the API.

The upcoming project `javaslang-pure` will cover pure functional programming. The classes are not lost, they will come back :-)

**Comment by lukaseder** [comment]
> ... *passed away*

:'-( ... ;)

Great. Looking forward to "pure", then!

**Comment by danieldietrich** [comment]
Wow, I closed this issue?! It was an accident :)

@lukaseder I liked the common Monad class.

> :'( ... ;)

I will defilibrate it today and see what is possible with `Kind`. It will stay with us until there is a better solution coming from the _pure_ module.

:)

**Comment by danieldietrich** [comment]
The patient did not survive. I have tried everything but Monad and Kind cannot be integrated. Locally I fiddled around hours with intersection types, changed the collection hierarchy, ... It simply does not work with Java or I don't see it.

I've to go further at this point.

The monadic types Option, Try, ..., List, ... have now a common type `TraversableOnce` which extends `Iterable`.

**Comment by danieldietrich** [comment]
@jbgi I had the idea to keep the javaslang-pure module very thin but 'deep'.

Cool would be a layer or some functionality, which _lifts_ arbitrary? types or at least javaslang types to algebraic types. Some kind of augmentation or decorator pattern. With this we can lift types into the pure core and operate algebraic on them...

I don't have a concrete solution, just a thought...

I'm on a short trip this weekend and mostly offline...

Have a great weekend!

\- Daniel

**Comment by talios** [comment]
> Everyone should have a Lukas :-)

Javaslang gets a new mascot icon?  :)

**Comment by danieldietrich** [comment]
Hehe :-) :+1:

**Comment by lukaseder** [comment]
> Javaslang gets a new mascot icon? :)

Let me know if you're interested. I can have someone send you our trademark licensing terms and prices.

**Comment by danieldietrich** [comment]
Then we will wait until Lukas is open sourced or available under a dual license.

**Comment by lukaseder** [comment]
You mean, there's potential of an upselling business in the mascot trade market? Intriguing...

**Comment by danieldietrich** [comment]
I will track functionaljava/functionaljava#126 to see how we can accomplish this in this context.

**Comment by danieldietrich** [comment]
Original topic:

> Add Apply between Monad and Functor

This is a duplicate now. With #803 we add Applicative (which is fused with Apply as known from Scalaz).
